apps/emqx_bridge/include/emqx_bridge_v2_register.hrl

```diff
@@ -0,0 +1,153 @@
+%% This function is called to register a bridge V2. It should be called before
+%% the system boots in a function triggered by an -on_load() directive
+%% since it should be called before the system boots because the config
+%% system depends on that.
+%%
+%% It is placed in an hrl file instead of in emqx_bridge_v2.erl because emqx_bridge_v2
+%% might not be loaded when the bridge module is loaded.
+-spec emqx_bridge_v2_register_bridge_type(#{
+    %% Should be provided by all bridges. Even if the bridge_v2_type_name is
+    %% the same as the bridge_v1_type_named.
+    'bridge_v1_type_name' := atom(),
+    'bridge_v2_type_name' := atom(),
+    'connector_type' := atom(),
+    'schema_module' := atom(),
+    'schema_struct_field' := atom() | binary()
+}) -> ok.
+emqx_bridge_v2_register_bridge_type(BridgeTypeInfo) ->
+    try
+        %% We must prevent overwriting so we take a lock when writing to persistent_term
+        global:set_lock(
+            {
+                internal_emqx_bridge_v2_persistent_term_info_key(),
+                internal_emqx_bridge_v2_persistent_term_info_key()
+            },
+            [node()],
+            infinity
+        ),
+        internal_maybe_create_initial_bridge_v2_info_map(),
+        internal_register_bridge_type_with_lock(BridgeTypeInfo)
+    catch
+        ErrorType:Reason:Stacktrace ->
+            %% Print the error on standard output as logger might not be
+            %% started yet
+            io:format("~p~n", [
+                #{
+                    'error_type' => ErrorType,
+                    'reason' => Reason,
+                    'stacktrace' => Stacktrace,
+                    'msg' => "Failed to register bridge V2 type"
+                }
+            ]),
+            erlang:raise(ErrorType, Reason, Stacktrace)
+    after
+        global:del_lock(
+            {
+                internal_emqx_bridge_v2_persistent_term_info_key(),
+                internal_emqx_bridge_v2_persistent_term_info_key()
+            },
+            [node()]
+        )
+    end,
+    ok.
+
+internal_register_bridge_type_with_lock(BridgeTypeInfo) ->
+    InfoMap0 = persistent_term:get(internal_emqx_bridge_v2_persistent_term_info_key()),
+    %% The Bridge V1 type is also a bridge V2 type due to backwards compatibility
+    InfoMap1 = emqx_utils_maps:deep_force_put(
+        [
+            bridge_v2_type_names,
+            maps:get(bridge_v1_type_name, BridgeTypeInfo)
```

Comment:
This could be absent for bridges created after this refactoring effort.

Reply:
The name can be the same, but yes it makes sense to let users register a bridge without providing a v1_type_name and which would mean that the v1 and v2 type names are the same.

Reply:
I mean that new bridges wouldn't have a V1 counterpart (or shouldn't, I think).

Reply:
Now, the V1 name has been made optional.